### PR TITLE
PLATFORM-2220: Automatically shorten memcached keys to be standard-compliant

### PR DESCRIPTION
**Pull request by wladekb**
Recently I found out that in a few cases we are trying to use memcached keys that are too long. This makes some of these data to be never cached.

This change introduces automatic and transparent key shortening by deriving a standard compliant key for the supplied one by cutting X first characters and appending a hash (sha1) of the entire key in such a way that the new key is never too long.

https://wikia-inc.atlassian.net/browse/PLATFORM-2220

/cc @owend @mixth-sense @macbre @kvas-damian

### CODE REVIEW

File: includes/objectcache/MemcachedClient.php [hunk 1]
```diff
@@ -1351,6 +1368,63 @@ function _flush_read_buffer( $f ) {
...
+	 * @param $rev_keys array Reverse key map obtained from call to _encode_keys()
+	 */
+	function _decode_keys( $result, $rev_keys ) {
+		$out = [];
+		foreach ($result as $key => $value) {
+			if (isset($rev_keys[$key])) {
+				$out[$rev_keys[$key]] = $value;
+			} else {
+				// @todo - it should not happen - log such cases
+				$out[$key] = $value;
```
Comment (macbre) [inline includes/objectcache/MemcachedClient.php:1427]: So let's log it :)

**Comment by macbre** [comment]
:+1:

File: lib/Wikia/src/Logger/WikiaLogger.php [hunk 2]
```diff
@@ -126,27 +126,38 @@ public function notice($message, Array $context=[]) {
...
 	}
 
 	public function log($level, $message, Array $context=[]) {} // NOOP
 
+	private function addStacktraceIfMissing( array &$context ) {
+		if ( !array_key_exists( 'exception', $context ) ) {
+			$context['exception'] = new \Exception();
+		}
+	}
+
```
Comment (macbre) [inline lib/Wikia/src/Logger/WikiaLogger.php:160]: :+1:

File: includes/objectcache/MemcachedClient.php [hunk 3]
```diff
@@ -1351,6 +1368,67 @@ function _flush_read_buffer( $f ) {
...
+			$this->_encode_key( $key[1] );
+
+			return;
+		}
+
+		$key_length = strlen( $key );
+		if ( $key_length > self::MEMCACHED_KEY_MAX_LENGTH ) {
+			$suffix = '--HASH--' . sha1( $key );
+			$suffix_length = strlen( $suffix );
+			$key = substr( $key, 0, self::MEMCACHED_KEY_MAX_LENGTH - $suffix_length ) . $suffix;
```
Comment (ghost) [inline includes/objectcache/MemcachedClient.php:1388]: Will that meet the 250 character protocol constraint? Plus, it is 250 characters or 250 bytes?

File: lib/Wikia/src/Logger/WikiaLogger.php (hunk 2 above)
Comment (ghost) [inline lib/Wikia/src/Logger/WikiaLogger.php:160]: Are you should it will not kill us in terms of Elasticsearch diskspace?

**Pull request merged**